core/src/main/java/org/apache/calcite/rel/type/TimeFrameSet.java

```diff
@@ -0,0 +1,762 @@
+        final BigFraction value2 = map2.get(entry.getKey());
+        if (value2 != null) {
+          return value2.divide(entry.getValue());
+        }
+      }
+      return null;
+    }
+
+    protected void expand(Map<TimeFrame, BigFraction> map, BigFraction f) {
+      map.put(this, f);
+    }
+
+    /** Adds a time frame like this to a builder. */
+    abstract void replicate(Builder b);
+
+    protected abstract CoreFrame core();
+
+    protected abstract BigFraction coreMultiplier();
+
+    @Override public boolean canRollUpTo(TimeFrame toFrame) {
+      if (toFrame == this) {
+        return true;
+      }
+      if (toFrame instanceof TimeFrameImpl) {
+        final TimeFrameImpl toFrame1 = (TimeFrameImpl) toFrame;
+        if (canDirectlyRollUp(this, toFrame1)) {
+          return true;
+        }
+        final TimeFrameSet frameSet = frameSet();
+        if (frameSet.rollupMap.entries().contains(Pair.of(this, toFrame1))) {
+          return true;
+        }
+        // Hard-code roll-up via DAY-to-MONTH bridge, for now.
+        final TimeFrameImpl day =
+            requireNonNull(frameSet.map.get(TimeUnit.DAY.name()));
+        final TimeFrameImpl month =
+            requireNonNull(frameSet.map.get(TimeUnit.MONTH.name()));
+        if (canDirectlyRollUp(this, day)
+            && canDirectlyRollUp(month, toFrame1)) {
+          return true;
+        }
+        // Hard-code roll-up via ISOWEEK-to-ISOYEAR bridge, for now.
+        final TimeFrameImpl isoYear =
+            requireNonNull(frameSet.map.get(TimeUnit.ISOYEAR.name()));
+        final TimeFrameImpl isoWeek =
+            requireNonNull(frameSet.map.get("ISOWEEK"));
+        if (canDirectlyRollUp(this, isoWeek)
+            && canDirectlyRollUp(isoYear, toFrame1)) {
+          return true;
+        }
+      }
+      return false;
+    }
+
+    private static boolean canDirectlyRollUp(TimeFrameImpl from,
+        TimeFrameImpl to) {
+      if (from.core().equals(to.core())) {
+        if (divisible(from.coreMultiplier(), to.coreMultiplier())) {
+          BigFraction diff = new BigFraction(from.core().epochDiff(from, to));
+          return divisible(from.coreMultiplier(), diff);
+        }
+        return false;
+      }
+      return false;
+    }
+
+    /** Returns whether {@code numerator} is divisible by {@code denominator}.
+     *
+     * <p>For example, {@code divisible(6, 2)} returns {@code true};
+     * {@code divisible(0, 2)} also returns {@code true};
+     * {@code divisible(2, 6)} returns {@code false}. */
+    private static boolean divisible(BigFraction numerator,
+        BigFraction denominator) {
+      return denominator.equals(BigFraction.ZERO)
+          || numerator
+          .divide(denominator)
+          .getNumerator()
+          .abs()
+          .equals(BigInteger.ONE);
+    }
+  }
+
+  /** Core time frame (such as SECOND, MONTH, ISOYEAR). */
+  static class CoreFrame extends TimeFrameImpl {
```

Comment:
Are there certain timeframes that couldn't be derived from those already in `TimeFrames.CORE`? 

Should all custom timeframes be a `SubFrame` of these known `CoreFrame`s? Perhaps I misread the builder but I wonder if people could end up making garbage `CoreFrame`s which could throw others sub and quotient ones off.

Reply:
Yes, in this feature we handle time frames that are multiples of the SECOND, MONTH, ISOWEEK/ISOYEAR core, possibly with an offset, or are aligned. 

Reasonable time frames that cannot be represented are Easter (years starting with Easter, and numbering days from then); and the Roman system (months divided into 8-9 day periods, kalends, ides etc.).

If those requirements arise, we can evolve the design.

I'll add a note to the Jira case.

Reply:
I'm sure people can define garbage frames if they try. The majority of people using this feature will be writing SQL against time frames that someone else has defined. Validation of time frames could be improved but is lower priority.